### PR TITLE
Extract MantaHttpRequestFactory from MantaConnectionFactory

### PR DESCRIPTION
**Pull request by tjcelaya**
This PR is the first part of a larger branch that works towards #310 and #311. This part extracts the convenience http request creation convenience methods from `MantaConnectionFactory` which is already responsible for configuring connection parameters and registering MBeans.

Apologies for the dull PR, most of the lines are replacements of `connectionFactory` -> `requestFactory`

### CODE REVIEW

**Comment by cburroughs** [comment]
I think the extraction looks good mechanically, but I fear this raises the specter from #316.  `MantaSeekableByteChannel`,  `EncryptionHttpHelper`, `MantaConnectionFactory` are all public classes whose public signatures are changed here. I'm not sure we ever came to a consensus on how to handle that.

**Comment by tjcelaya** [comment]
Pushed a few commits to actually deprecate the old constructors instead of just dropping them entirely.

*Note*: the `ConnectionContextAware` interface causes the visibility of `StandardHttpHelper#getConnectionContext` to change from `protected` to `public`. I don't see a way to maintain the previous visibility level in combination with these changes but I also don't think it's too much of an issue.

**Comment by tjcelaya** [comment]
# Final summary of Changes

- `MantaClient` lost references to `MantaConnectionFactory` and `MantaConnectionContext` since those are used primarily by `HttpHelper`
- `MantaClient#close` now only needs to worry about `danglingStreams` and `httpHelper`
- `HttpHelper` now has a method for retrieving its `MantaHttpRequestFactory` since Multipart managers have a habit of building raw request objects. This is in addition to the existing method for grabbing a connection context
- The responsibility for closing `MantaConnectionContext` has been moved to `HttpHelper`
- `MantaConnectionContext` (or at least the only implementor) is now responsible for closing both the factory and the client. This makes sense since it manages bot the pool owner (connectionFactory) and the pool user (httpClient)

## Old close chain
```
MantaClient --> HttpHelper             -> MantaConnectionContext -> HttpClient
            \-> MantaConnectionFactory
```
## New close chain
```
MantaClient -> HttpHelper             -> MantaConnectionContext --> HttpClient
                                                                \-> MantaConnectionFactory
```

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/JobsMultipartManager.java [hunk 1]
```diff
@@ -303,12 +294,12 @@ MantaMultipartUploadPart uploadPart(final JobsMultipartUpload upload,
...
 
         final String path = multipartPath(upload.getId(), partNumber);
-        final HttpPut put = connectionFactory.put(path);
+        final HttpPut put = httpHelper.getRequestFactory().put(path);
         put.setEntity(entity);
 
         final int expectedStatusCode = HttpStatus.SC_NO_CONTENT;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(put, context)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(put, context)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/JobsMultipartManager.java]: This is a long chain of linked objects. Maybe break out the context or client as its own variable?

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 2]
```diff
@@ -264,7 +252,7 @@ public ServerSideMultipartUpload initiateUpload(final String path,
 
         final int expectedStatusCode = HttpStatus.SC_CREATED;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(post)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(post)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) blah blah...

We may want to break this (context or client) out to a separate variable to make it easier to read.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 3]
```diff
@@ -324,10 +312,10 @@ MantaMultipartUploadPart uploadPart(final ServerSideMultipartUpload upload,
         final int adjustedPartNumber = partNumber - 1;
 
         final String putPath = upload.getPartsDirectory() + SEPARATOR + adjustedPartNumber;
-        final HttpPut put = connectionFactory.put(putPath);
+        final HttpPut put = httpHelper.getRequestFactory().put(putPath);
         put.setEntity(entity);
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(put, context)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(put, context)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 4]
```diff
@@ -360,13 +348,13 @@ public MantaMultipartUploadPart getPart(final ServerSideMultipartUpload upload,
...
         final String getPath = upload.getPartsDirectory() + SEPARATOR + "state";
-        final HttpGet get = connectionFactory.get(getPath);
+        final HttpGet get = httpHelper.getRequestFactory().get(getPath);
 
         final String objectPath;
 
         final int expectedStatusCode = HttpStatus.SC_OK;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(get)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(get)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 5]
```diff
@@ -394,11 +382,11 @@ public MantaMultipartUploadPart getPart(final ServerSideMultipartUpload upload,
         }
 
         final String headPath = upload.getPartsDirectory() + SEPARATOR + adjustedPartNumber;
-        final HttpHead head = connectionFactory.head(headPath);
+        final HttpHead head = httpHelper.getRequestFactory().head(headPath);
 
         final String etag;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(head)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(head)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 6]
```diff
@@ -439,11 +427,11 @@ public MantaMultipartStatus getStatus(final ServerSideMultipartUpload upload)
         }
 
         final String getPath = partsDirectory + SEPARATOR + "state";
-        final HttpGet get = connectionFactory.get(getPath);
+        final HttpGet get = httpHelper.getRequestFactory().get(getPath);
 
         final int expectedStatusCode = HttpStatus.SC_OK;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(get)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(get)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 7]
```diff
@@ -543,11 +531,11 @@ public void abort(final ServerSideMultipartUpload upload) throws IOException {
         Validate.notNull(upload, "Upload state object must not be null");
 
         final String postPath = upload.getPartsDirectory() + SEPARATOR + "abort";
-        final HttpPost post = connectionFactory.post(postPath);
+        final HttpPost post = httpHelper.getRequestFactory().post(postPath);
 
         final int expectedStatusCode = HttpStatus.SC_NO_CONTENT;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(post)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(post)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java [hunk 8]
```diff
@@ -617,7 +605,7 @@ public void complete(final ServerSideMultipartUpload upload,
 
         final int expectedStatusCode = HttpStatus.SC_CREATED;
 
-        try (CloseableHttpResponse response = connectionContext.getHttpClient().execute(post)) {
+        try (CloseableHttpResponse response = httpHelper.getConnectionContext().getHttpClient().execute(post)) {
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/multipart/ServerSideMultipartManager.java]: Potentially break context or client out as a separate variable.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaSeekableByteChannel.java [hunk 9]
```diff
@@ -135,7 +132,56 @@ protected MantaSeekableByteChannel(final AtomicReference<HttpUriRequest> request
         this.responseStream = responseStream;
         this.path = path;
         this.position = position;
-        this.connectionFactory = connectionFactory;
+        this.httpHelper = httpHelper;
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaSeekableByteChannel.java]: Can you checkout the Manta Hadoop Filesystem Driver and try to refactor it to the non-deprecated API as a sanity check?

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/http/EncryptionHttpHelper.java [hunk 10]
```diff
@@ -110,17 +110,29 @@
...
 
+
+    /**
+     * Creates a new instance of the helper class.
+     *  @param connectionContext saved context used between requests to the Manta client
+     * @param config configuration context object
+     */
+    public EncryptionHttpHelper(final MantaConnectionContext connectionContext,
+                                final ConfigContext config) {
+        this(connectionContext, null, config);
```
Comment (dekobon) [inline java-manta-client-unshaded/src/main/java/com/joyent/manta/http/EncryptionHttpHelper.java]: Don't make the non-deprecated constructor call out to the deprecated constructor. Create a new constructor that doesn't use the deprecated code path.

Comment (dekobon) [reply java-manta-client-unshaded/src/main/java/com/joyent/manta/http/EncryptionHttpHelper.java]: This is the only substantial comment I had on the changes and the only reason I asked for changes.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaSeekableByteChannel.java (hunk 9 above)
Comment (tjcelaya) [reply java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaSeekableByteChannel.java]: Checkout out hadoop-manta but it never touches the `MantaSeekableByteChannel` constructor. All access to this class is through `MantaClient#getSeekableByteChannel(String)` and `MantaSeekableByteChannel#position(long)`.

**Pull request merged**